compiler/rustc_resolve/src/late.rs

```diff
@@ -4150,34 +4150,36 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             PathResult::Indeterminate => bug!("indeterminate path result in resolve_qpath"),
         };
 
-        if path.len() > 1
-            && let Some(res) = result.full_res()
-            && let Some((&last_segment, prev_segs)) = path.split_last()
-            && prev_segs.iter().all(|seg| !seg.has_generic_args)
-            && res != Res::Err
-            && path[0].ident.name != kw::PathRoot
-            && path[0].ident.name != kw::DollarCrate
-        {
-            let unqualified_result = {
-                match self.resolve_path(&[last_segment], Some(ns), None) {
-                    PathResult::NonModule(path_res) => path_res.expect_full_res(),
-                    PathResult::Module(ModuleOrUniformRoot::Module(module)) => {
-                        module.res().unwrap()
-                    }
-                    _ => return Ok(Some(result)),
-                }
-            };
-            if res == unqualified_result {
-                let lint = lint::builtin::UNUSED_QUALIFICATIONS;
+        let end_pos =
+            path.iter().position(|seg| seg.has_generic_args).map_or(path.len(), |pos| pos + 1);
+        if path.iter().all(|seg| !seg.ident.span.from_expansion()) {
+            let unqualified =
+                path[..end_pos].iter().enumerate().skip(1).rev().find_map(|(i, seg)| {
```

Comment:
`path[..end_pos].iter()` is the same as `path.iter().take_while(|seg| !seg.has_generic_args)`, right? Then we don't need the position.

Reply:
Logically it's that plus one segment, `itertools` has a `take_while_inclusive` but neither are double ended

Reply:
Ok, then `end_pos` can be just moved into the `if` condition.